src/semaphore.cr

```diff
@@ -0,0 +1,89 @@
+require "./dll"
+require "./safe"
+require "./waiter"
+
+module Sync
+  # Counting semaphore.
+  #
+  # Allows a limited number of fibers to enter a critical section at a time.
+  #
+  # A counting semaphore can be a simple alternative to starting multiple
+  # fibers, and requiring a `Channel` to pass work to execute, and maybe a
+  # `Future` to get a value back. The semaphore is simpler, but starting
+  # explicit fibers might prove better for your application, depending on your
+  # workload.
+  #
+  # In the following example, only 4 fibers can compute a password hash
+  # concurrently. If the computer has more than 4 CPU cores then only 4 threads
+  # may be blocked hashing a password, while the other threads will keep running
+  # the other fibers.
+  #
+  # ```
+  # LOGINS = Sync::Semaphore.new(4)
+  #
+  # def hash(password)
+  #   LOGINS.synchronize do
+  #     Crypto::Bcrypt::Password.create("super secret", cost: 10)
+  #   end
+  # end
+  # ```
+  #
+  # NOTE: A semaphore should have a value stricly greater than 1. Prefer a
+  # `Mutex` if the concurrency of the critical section must be limited to one
+  # exclusive fiber.
+  @[Sync::Safe]
+  struct Semaphore
+    def initialize(value : Int32)
+      @value = Atomic(Int32).new(value)
+      @mu = MU.new
+      @waiters = Dll(Waiter).new
```

Comment:
We don't need a `MU` to protect the `Dll`: there's no concurrency issue, there's only a parallelism issue. It should use a `Thread::Mutex` (or a `MU` but designed for threads).